src/main/java/com/conveyal/datatools/manager/controllers/api/LabelController.java

```diff
@@ -0,0 +1,234 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.conveyal.datatools.common.utils.SparkUtils.getPOJOFromRequestBody;
+import static com.conveyal.datatools.common.utils.SparkUtils.logMessageAndHalt;
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Filters.in;
+import static spark.Spark.delete;
+import static spark.Spark.get;
+import static spark.Spark.post;
+import static spark.Spark.put;
+
+
+public class LabelController {
+    private static final Logger LOG = LoggerFactory.getLogger(LabelController.class);
+    private static JsonManager<Label> json = new JsonManager<>(Label.class, JsonViews.UserInterface.class);
+
+
+    /**
+     * Spark HTTP endpoint to get a single label by ID
+     */
+    private static Label getLabel(Request req, Response res) {
+        return requestLabelById(req, Actions.VIEW);
+    }
+
+    /**
+     * Spark HTTP endpoint that handles getting all labels for a handful of use cases:
+     * - for a single project (if projectId query param provided)
+     */
+    private static Collection<Label> getAllLabels(Request req, Response res) {
+        Auth0UserProfile user = req.attribute("user");
+        String projectId = req.queryParams("projectId");
+        Project project = Persistence.projects.getById(projectId);
+        if (project == null) {
+            logMessageAndHalt(req, 400, "Must provide valid projectId query param to retrieve labels.");
+        }
+
+        boolean isProjectAdmin = user.canAdministerProject(project.id, project.organizationId);
+
+        return project.retrieveProjectLabels(isProjectAdmin);
+    }
+
+    /**
+     * HTTP endpoint to create a new label
+     */
+    private static Label createLabel(Request req, Response res) throws IOException {
+        Auth0UserProfile userProfile = req.attribute("user");
+        Label newLabel = getPOJOFromRequestBody(req, Label.class);
+        validate(req, newLabel);
+        // User may not be allowed to create a new label
+        checkLabelPermissions(req, newLabel, Actions.CREATE);
+
+        try {
+            Persistence.labels.create(newLabel);
+            return newLabel;
+        } catch (Exception e) {
+            logMessageAndHalt(req, 500, "Unknown error encountered creating label.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Check that updated or new label object is valid. This method should be called before a label is
+     * persisted to the database.
+     */
+    private static void validate(Request req, Label label) {
+        // Label is quite forgiving (sets defaults if null) and the boolean value is type checked,
+        // so there is little to validate.
+        if (label.name.length() > 25) {
+            logMessageAndHalt(
+                    req,
+                    HttpStatus.BAD_REQUEST_400,
+                    "Request was invalid, the name was too long."
```

Comment:
Please improve this error message to say what the maximum length is.

Reply:
Added in 4ac2e4cab7a40d0c40960ba15503c2781adef26d